### PR TITLE
add ipfs binding

### PR DESCRIPTION
**Pull request by hypernova3**
# Description

dapr meets dapps 😀🚀

new binding `ipfs` for IPFS. can be used to build dapps and work with things like web3 and NFT.

supports get, add, pin-add, pin-ls, delete (=pin-rm). similar to IPFS cli.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation / Created issue in the https://github.com/dapr/docs/pull/2661

### CODE REVIEW

**Comment by daixiang0** [comment]
@hypernova3 thanks for your contributions!

Could you provide a link about ipfs in the PR description? 

Also, is it used in your company or maintained by you? Just want to ensure new implementation is widely used rather than just implemented.

**Comment by hypernova3** [comment]
@daixiang0 IPFS: https://ipfs.io/ many people use it for building web3 dapps or store NFTs.
https://www.one37pm.com/nft/what-is-ipfs

i do not work for protocol labs. 

this binding can be used with the public IPFS network or can create a private network.

File: go.sum [hunk 1]
```diff
@@ -67,15 +70,25 @@ cloud.google.com/go/storage v1.10.0 h1:STgFzyU5/8miMl0//zKh2aQeTyeaUH3WN9bSUiJ09
 cloud.google.com/go/storage v1.10.0/go.mod h1:FLPqc6j+Ki4BU591ie1oL6qBQGu2Bl/tZ9ullr3+Kg0=
 contrib.go.opencensus.io/exporter/prometheus v0.4.1 h1:oObVeKo2NxpdF/fIfrPsNj6K0Prg0R0mHM+uANlYMiM=
 contrib.go.opencensus.io/exporter/prometheus v0.4.1/go.mod h1:t9wvfitlUjGXG2IXAZsuFq26mDGid/JwCEXp+gTG/9U=
+dmitri.shuralyov.com/app/changes v0.0.0-20180602232624-0a106ad413e3/go.mod h1:Yl+fi1br7+Rr3LqpNJf1/uxUdtRUV+Tnj0o93V2B9MU=
```
Comment (berndverst) [inline go.sum:73]: This is missing a LICENSE. As such we cannot accept this dependency.

File: go.sum [hunk 2]
```diff
@@ -67,15 +70,25 @@ cloud.google.com/go/storage v1.10.0 h1:STgFzyU5/8miMl0//zKh2aQeTyeaUH3WN9bSUiJ09
...
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
+dmitri.shuralyov.com/gpu/mtl v0.0.0-20201218220906-28db891af037/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
+dmitri.shuralyov.com/html/belt v0.0.0-20180602232347-f7d459c86be0/go.mod h1:JLBrvjyP0v+ecvNYvCpyZgu5/xkfAUhi6wJj28eUfSU=
```
Comment (berndverst) [inline go.sum:76]: This is missing a LICENSE as such we cannot accept this dependency.

**Review by berndverst — CHANGES_REQUESTED** [review]
Please ensure to use dependencies with appropriate licenses.

File: go.sum (hunk 1 above)
Comment (ItalyPaleAle) [reply go.sum:73]: False positive: https://dmitri.shuralyov.com/app/changes$file/LICENSE MIT-licensed

**Comment by hypernova3** [comment]
added docs: https://github.com/dapr/components-contrib/pull/1849

PR is ready. please review

**Review by berndverst — CHANGES_REQUESTED** [review]
Thanks for the contributions, this looks great.

Can you please move this binding to `bindings/ipfs` and drop the `web3` from the folder structure and name?

Similar to the `http` binding, this can simply be the `ipfs` binding.

Thanks!

**Comment by berndverst** [comment]
This must not be merged until https://github.com/dapr/dapr/pull/4808 is resolved.

**Comment by hypernova3** [comment]
@berndverst changed to `bindings.ipfs`. let's wait for https://github.com/dapr/dapr/pull/4808

**Review by berndverst — APPROVED** [review]

**Pull request merged**